src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1354,13 +1363,49 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
             try
             {
                 // Send headers
-                http2Stream = await SendHeadersAsync(request, cancellationToken).ConfigureAwait(false);
-
-                // Send request body, if any
-                await http2Stream.SendRequestBodyAsync(cancellationToken).ConfigureAwait(false);
+                bool shouldExpectContinue = request.Content != null && request.HasHeaders && request.Headers.ExpectContinue == true;
 
-                // Wait for response headers to be read.
-                await http2Stream.ReadResponseHeadersAsync().ConfigureAwait(false);
+                http2Stream = await SendHeadersAsync(request, cancellationToken, mustFlush: shouldExpectContinue).ConfigureAwait(false);
+                if (shouldExpectContinue)
+                {
+                    // Send header and wait a little bit to see if server sends 100, reject code or nothing.
+                    if (NetEventSource.IsEnabled) Trace($"Request content is not null, start processing 100-Continue.");
+                    await http2Stream.SendRequestBodyWithExpect100ContinueAsync(cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    // Send request body, if any
+                    Task bodyTask = http2Stream.SendRequestBodyAsync(cancellationToken);
+                    // read response headers.
+                    Task responseHeadersTask = http2Stream.ReadResponseHeadersAsync();
+                    Task finished = await Task.WhenAny(new Task[] { bodyTask, responseHeadersTask}).ConfigureAwait(false);
+                    if (finished == bodyTask)
+                    {
+                        // Sending request body finished before getting headers.
+                        Task t = bodyTask;
+                        bodyTask = null;
+                        await t.ConfigureAwait(false);
+                        // Wait for response headers again.
+                        await responseHeadersTask.ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        if (bodyTask.IsCompleted)
```

Comment:
If we get here, the responseHeadersTask completed, right?  We're not awaiting it to propagate any exceptions that may have occurred.

Seems like this logic could be restructured a little, e.g.
```C#
if (bodyTask == await Task.WhenAny(bodyTask, responseHeadersTask).ConfigureAwait(false) ||
    bodyTask.IsCompleted)
{
    // The sending of the request body completed before receiving all of the request headers.
    Task t = bodyTask;
    bodyTask = null;
    try
    {
        await t.ConfigureAwait(false);
    }
    catch
    {
        LogExceptionAsync(this, responseHeadersTask, "Response headers"); // ContinueWith logic factored into a helper
        throw;
    }

    await responseHeadersTask.ConfigureAwait(false);
}
else
{
    // We received the response headers but the request body hasn't yet finished.
    bodyTask.LogExceptionAsync(this, responseHeadersTask, "Request body");
}
```
That said, I'm not clear on the `else` case above where the request headers have completed but the response body hasn't.  If the response body subsequently fails and we're still handling the response body, wouldn't we want to propagate any failure there out via the response body handling?  If that's already completed as well, we're out of luck and logging is the best we can do, but if it hasn't yet completed, seems we'd want to take the opportunity to forward the error out?

Reply:
we don't have problem with ProcessIncomingFramesAsync() because we catch all exceptions, right? If so, can we do something similar and use _abortException?

Reply:
BTW we have few places where we have "ignored" task. Do we have same issue there?

Reply:
> BTW we have few places where we have "ignored" task. Do we have same issue there?

If exceptions can go unhandled in their implementations, then yes.

Reply:
> we don't have problem with ProcessIncomingFramesAsync() because we catch all exceptions, right?

Right.

> If so, can we do something similar and use _abortException?

Quite possibly.

Reply:
I updated the code. I did look at what is happening with SendRequestBodyAsync(). If wee get IO exception or if server sends something we do not like (include RST and GOAWAY) we will store that in stream._abortException and we will throw if we still processing response. 
I could explore that more as followup (so as ignored tasks) and we could possibly save all exceptions from SendRequestBodyAsync(). We could also log everything in there to simplify the ContinueWith()

Reply:
> I could explore that more as followup (so as ignored tasks) and we could possibly save all exceptions from SendRequestBodyAsync()

Yes, let's please follow-up.  There are two concerns:
1. That if an error occurs from sending we don't lose the exception if at all possible.  That means best case propagating it to the response and worst case logging it.
2. That we don't end up with Faulted Tasks that never have their Exceptions observed, which will cause noise for anyone listening to TaskScheduler.UnobservedTaskExceptions.